File: internal/model/monitor_config.go
package model

import (
	"encoding/json"
	"time"
)

// MonitorScope / ScopeHeld / ScopeCustom 已在 subscription.go 中定义，此处复用

// ============================================================================
//  MonitorConfig 盯盘监控配置
// ============================================================================

// MonitorConfig 盯盘监控配置
type MonitorConfig struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	UID       uint           `gorm:"index;not null" json:"uid"`
	Name      string         `gorm:"size:100;not null" json:"name"`
	Scope     MonitorScope   `gorm:"size:20;not null;default:held" json:"scope"`
	Stocks    string         `gorm:"type:text" json:"stocks"` // JSON: ["000001","600036"]
	Rules     string         `gorm:"type:text;not null" json:"rules"`
	Cooldown  string         `gorm:"type:text;not null" json:"cooldown"`
	Template  string         `gorm:"type:text" json:"template"`
	IsActive  bool           `gorm:"default:true;index" json:"is_active"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt *time.Time     `gorm:"index" json:"-"`
}

func (MonitorConfig) TableName() string { return "monitor_configs" }

// ============================================================================
//  MonitorConfigBot 监控配置-机器人关联
// ============================================================================

// MonitorConfigBot 监控配置关联机器人
type MonitorConfigBot struct {
	ID        uint      `gorm:"primarykey" json:"id"`
	ConfigID  uint      `gorm:"uniqueIndex:idx_config_bot" json:"config_id"`
	BotID     uint      `gorm:"uniqueIndex:idx_config_bot" json:"bot_id"`
	CreatedAt time.Time `json:"created_at"`
}

func (MonitorConfigBot) TableName() string { return "monitor_config_bots" }

// ============================================================================
//  告警规则类型（动态，每种类型有独立参数）
// ============================================================================

// RuleType 告警规则类型
type RuleType string

const (
	RuleTypeDailyChange  RuleType = "daily_change"  // 当日涨幅（6档）
	RuleTypeRapidMove    RuleType = "rapid_move"     // 急拉急跌
	RuleTypeVolumeRatio  RuleType = "volume_ratio"   // 量比异动
	RuleTypeSealBoard    RuleType = "seal_board"     // 封单数量

	// 兼容旧版
	_legacyShortTermVolatility = "short_term_volatility"

	// 兼容旧版（已废弃，保留枚举值便于解析错误提示）
	_legacyLimitUp      = "limit_up"
	_legacyLimitDown    = "limit_down"
	_legacySurgeUp      = "surge_up"
	_legacySurgeDown    = "surge_down"
	_legacyHighTurnover = "high_turnover"
)

// MonitorRule 单条告警规则（type + params）
type MonitorRule struct {
	Type   RuleType        `json:"type"`
	Params json.RawMessage `json:"params"` // 类型特定的参数 JSON
}

// ============================================================================
//  各规则类型的参数结构体
// ============================================================================

// DailyChangeParams 当日涨幅参数
type DailyChangeParams struct {
	SurgeBigEnabled   bool    `json:"surge_big_enabled"`   // 大涨告警开关
	SurgeBig          float64 `json:"surge_big"`           // 大涨阈值(%)
	SurgeSmallEnabled bool    `json:"surge_small_enabled"` // 小涨告警开关
	SurgeSmall        float64 `json:"surge_small"`         // 小涨阈值(%)
	LimitUpEnabled    bool    `json:"limit_up_enabled"`    // 涨停告警开关
	LimitUp           float64 `json:"limit_up"`            // 涨停阈值(%)
	LimitDownEnabled  bool    `json:"limit_down_enabled"`  // 跌停告警开关
	LimitDown         float64 `json:"limit_down"`          // 跌停阈值(%)
	DropSmallEnabled  bool    `json:"drop_small_enabled"`  // 小跌告警开关
	DropSmall         float64 `json:"drop_small"`          // 小跌阈值(%)
	DropBigEnabled    bool    `json:"drop_big_enabled"`    // 大跌告警开关
	DropBig           float64 `json:"drop_big"`            // 大跌阈值(%)
}

// DefaultDailyChangeParams 默认当日涨幅参数（全部启用）
func DefaultDailyChangeParams() DailyChangeParams {
	return DailyChangeParams{
		SurgeBigEnabled:   true,
		SurgeBig:          9.0,
		SurgeSmallEnabled: true,
		SurgeSmall:        5.0,
		LimitUpEnabled:    true,
		LimitUp:           9.8,
		LimitDownEnabled:  true,
		LimitDown:         -9.8,
		DropSmallEnabled:  true,
		DropSmall:         -5.0,
		DropBigEnabled:    true,
		DropBig:           -9.0,
	}
}

// RapidMoveParams 急拉急跌参数
type RapidMoveParams struct {
	Minutes      int     `json:"minutes"`       // 时间窗口(分钟)
	AmplitudePct float64 `json:"amplitude_pct"` // 涨跌幅阈值(%)
	UpEnabled    bool    `json:"up_enabled"`    // 上涨方向启用
	DownEnabled  bool    `json:"down_enabled"`  // 下跌方向启用
}

// DefaultRapidMoveParams 默认急拉急跌参数
func DefaultRapidMoveParams() RapidMoveParams {
	return RapidMoveParams{Minutes: 5, AmplitudePct: 3.0, UpEnabled: true, DownEnabled: true}
}

// VolumeRatioParams 量比异动参数
type VolumeRatioParams struct {
	MinRatio float64 `json:"min_ratio"` // 量比阈值（当日量/5日均量）
}

// DefaultVolumeRatioParams 默认量比参数
func DefaultVolumeRatioParams() VolumeRatioParams {
	return VolumeRatioParams{MinRatio: 3.0}
}

// SealBoardParams 涨跌停封单数量监控参数
//
// 涨停时监控买一(bid)，跌停时监控卖一(ask)，共用同一阈值
type SealBoardParams struct {
	MinLots int `json:"min_lots"` // 封单小于此手数时告警（涨停看买一，跌停看卖一）
}

// DefaultSealBoardParams 默认封单参数
func DefaultSealBoardParams() SealBoardParams {
	return SealBoardParams{MinLots: 1000}
}

// ============================================================================
//  规则类型中文标签
// ============================================================================

// RuleTypeLabel 规则类型中文标签
var RuleTypeLabel = map[RuleType]string{
	RuleTypeDailyChange: "当日涨幅监控",
	RuleTypeRapidMove:   "急拉急跌监控",
	RuleTypeVolumeRatio: "量比异动监控",
	RuleTypeSealBoard:   "涨跌停封单监控",
}

// DailyChangeSubLabel 当日涨幅子标签
var DailyChangeSubLabel = map[string]string{
	"surge_big":   "大涨",
	"surge_small": "小涨",
	"limit_up":    "涨停",
	"limit_down":  "跌停",
	"drop_small":  "小跌",
	"drop_big":    "大跌",
}

// ============================================================================
//  冷却配置
// ============================================================================

// MonitorCooldown 冷却策略
type MonitorCooldown struct {
	IntervalMinutes int `json:"interval_minutes"` // 同股同类型告警最小间隔(分钟)
	DailyMax        int `json:"daily_max"`        // 每天同股同类型最多推送次数
}

// ============================================================================
//  Parse 方法
// ============================================================================

// ParseRule 解析 Rule JSON
func (c *MonitorConfig) ParseRule() (*MonitorRule, error) {
	if c.Rules == "" {
		return nil, nil
	}
	var rule MonitorRule
	if err := json.Unmarshal([]byte(c.Rules), &rule); err != nil {
		return nil, err
	}
	if rule.Type == "" {
		return nil, nil
	}
	return &rule, nil
}

// ParseCooldown 解析 Cooldown JSON
func (c *MonitorConfig) ParseCooldown() (*MonitorCooldown, error) {
	var cd MonitorCooldown
	if err := json.Unmarshal([]byte(c.Cooldown), &cd); err != nil {
		return nil, err
	}
	return &cd, nil
}

// ParseStocks 解析 Stocks JSON
func (c *MonitorConfig) ParseStocks() ([]string, error) {
	var stocks []string
	if c.Stocks == "" {
		return nil, nil
	}
	if err := json.Unmarshal([]byte(c.Stocks), &stocks); err != nil {
		return nil, err
	}
	return stocks, nil
}

// ============================================================================
//  MonitorConfigDetail 响应 DTO（含关联机器人名称）
// ============================================================================

// MonitorConfigDetail 监控配置详情（含关联机器人）
type MonitorConfigDetail struct {
	ID        uint              `json:"id"`
	UID       uint              `json:"uid"`
	Name      string            `json:"name"`
	Scope     MonitorScope      `json:"scope"`
	Stocks    []string          `json:"stocks"`
	Rule      MonitorRule      `json:"rule"`
	Cooldown  *MonitorCooldown  `json:"cooldown"`
	Template  string            `json:"template"`
	IsActive  bool              `json:"is_active"`
	Bots      []PushBot         `json:"bots"`
	CreatedAt time.Time         `json:"created_at"`
	UpdatedAt time.Time         `json:"updated_at"`
}

File: web/src/api/monitor-config.ts
import { getToken } from '../utils/auth'

const BASE = 'http://localhost:9100/api/v1/monitor-configs'

// ========== 请求封装 ==========

function request(url: string, options?: RequestInit): Promise<any> {
  const token = getToken()
  return fetch(url, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...((options?.headers as Record<string, string>) || {}),
    },
  }).then(async (res) => {
    const data = await res.json()
    if (!res.ok) throw new Error(data.message || '请求失败')
    return data
  })
}

// ========== 类型定义 ==========

/** 监控范围 */
export type MonitorScope = 'held' | 'custom'

/** 规则类型 */
export type RuleType =
  | 'daily_change'
  | 'rapid_move'
  | 'volume_ratio'
  | 'seal_board'

/** 告警规则（动态类型 + 参数） */
export interface MonitorRule {
  type: RuleType
  params: Record<string, number | boolean> // 类型特定参数
}

/** 当日涨幅参数 */
export interface DailyChangeParams {
  surge_big_enabled: boolean
  surge_big: number
  surge_small_enabled: boolean
  surge_small: number
  limit_up_enabled: boolean
  limit_up: number
  limit_down_enabled: boolean
  limit_down: number
  drop_small_enabled: boolean
  drop_small: number
  drop_big_enabled: boolean
  drop_big: number
}

/** 急拉急跌参数 */
export interface RapidMoveParams {
  minutes: number
  amplitude_pct: number
  up_enabled: boolean
  down_enabled: boolean
}

/** 量比异动参数 */
export interface VolumeRatioParams {
  min_ratio: number
}

/** 封单监控参数 */
export interface SealBoardParams {
  min_lots: number
}

/** 冷却策略 */
export interface MonitorCooldown {
  interval_minutes: number
  daily_max: number
}

/** 机器人信息 */
export interface BotInfo {
  id: number
  name: string
  channel: string
}

/** 监控配置详情 */
export interface MonitorConfigDetail {
  id: number
  uid: number
  name: string
  scope: MonitorScope
  stocks: string[]
  rule: MonitorRule
  cooldown: MonitorCooldown
  template: string
  is_active: boolean
  bots: BotInfo[]
  created_at: string
  updated_at: string
}

/** 创建监控配置请求 */
export interface CreateMonitorConfigReq {
  name: string
  scope?: MonitorScope
  stocks?: string[]
  rule: MonitorRule
  cooldown: MonitorCooldown
  template?: string
  bot_ids?: number[]
}

/** 更新监控配置请求（部分字段） */
export interface UpdateMonitorConfigReq {
  name: string
  scope: MonitorScope
  stocks?: string[]
  rule: MonitorRule
  cooldown: MonitorCooldown
  template: string
}

/** 列表响应 */
export interface MonitorConfigListResp {
  data: MonitorConfigDetail[]
  total: number
  page: number
  page_size: number
}

// ========== 常量映射 ==========

export const ScopeLabels: Record<MonitorScope, string> = {
  held: '我的持仓',
  custom: '自选股票',
}

export const RuleTypeLabels: Record<RuleType, string> = {
  daily_change: '当日涨幅监控',
  rapid_move: '急拉急跌监控',
  volume_ratio: '量比异动监控',
  seal_board: '涨跌停封单监控',
}

/** 当日涨幅子类型 */
export const DailyChangeSubKeys = [
  { key: 'surge_big', label: '大涨阈值(%)' },
  { key: 'surge_small', label: '小涨阈值(%)' },
  { key: 'limit_up', label: '涨停阈值(%)' },
  { key: 'limit_down', label: '跌停阈值(%)' },
  { key: 'drop_small', label: '小跌阈值(%)' },
  { key: 'drop_big', label: '大跌阈值(%)' },
] as const

/** 短期振幅参数 key */
export const VolatilityParamKeys = [
  { key: 'minutes', label: '时间窗口(分钟)' },
  { key: 'amplitude_pct', label: '振幅阈值(%)' },
] as const

/** 封单参数 key */
export const SealBoardParamKeys = [
  { key: 'min_lots', label: '封单小于(手)' },
] as const

export const DefaultCooldown: MonitorCooldown = {
  interval_minutes: 5,
  daily_max: 3,
}

/** 每种规则类型的默认推送文案 */
export const DefaultRuleTemplates: Record<RuleType, string> = {
  daily_change: '[${alert_label}] ${name}(${code}) 报 ${price} (${change_pct}%)',
  rapid_move: '[${alert_label}] ${name}(${code}) ${change_pct}% 报 ${price}',
  volume_ratio: '[${alert_label}] ${name}(${code}) 量比 ${ratio} 成交 ${volume}',
  seal_board: '[${alert_label}] ${name}(${code}) 封单不足 ${min_lots}手',
}

/** 所有可用模板变量及其含义 */
export const AllTemplateVars = {
  name: '股票名称',
  code: '股票代码',
  alert_type: '规则大类 (daily_change / rapid_move / ...)',
  alert_subtype: '告警子类型 (surge_big / rapid_up / ...)',
  alert_label: '告警中文标签 (涨停 / 急拉 / 量比异动 / ...)',
  price: '当前价格(元)',
  change_pct: '涨跌幅 %',
  turnover: '换手率 %',
  volume: '成交量',
  minutes: '时间窗口(分钟) — 仅急拉急跌',
  amplitude: '幅度阈值 — 仅急拉急跌',
  ratio: '计算出的实际量比 — 仅量比异动',
  min_lots: '封单阈值(手) — 仅封单监控',
} as const

// ========== API 函数 ==========

export function listMonitorConfigs(
  page = 1,
  pageSize = 20,
): Promise<MonitorConfigListResp> {
  const params = new URLSearchParams({ page: String(page), page_size: String(pageSize) })
  return request(`${BASE}?${params}`)
}

export function getMonitorConfig(id: number): Promise<MonitorConfigDetail> {
  return request(`${BASE}/${id}`).then((d) => d.data)
}

export function createMonitorConfig(data: CreateMonitorConfigReq): Promise<MonitorConfigDetail> {
  return request(BASE, { method: 'POST', body: JSON.stringify(data) }).then((d) => d.data)
}

export function updateMonitorConfig(id: number, data: UpdateMonitorConfigReq): Promise<MonitorConfigDetail> {
  return request(`${BASE}/${id}`, { method: 'PUT', body: JSON.stringify(data) }).then((d) => d.data)
}

export function deleteMonitorConfig(id: number): Promise<void> {
  return request(`${BASE}/${id}`, { method: 'DELETE' })
}

export function setMonitorConfigActive(id: number, active: boolean): Promise<void> {
  return request(`${BASE}/${id}/active`, {
    method: 'PATCH',
    body: JSON.stringify({ is_active: active }),
  })
}

export function updateMonitorConfigBots(id: number, botIDs: number[]): Promise<void> {
  return request(`${BASE}/${id}/bots`, {
    method: 'PUT',
    body: JSON.stringify({ bot_ids: botIDs }),
  })
}
